toolchain/check/cpp_thunk.cpp

```diff
@@ -0,0 +1,369 @@
+
+  for (const clang::ParmVarDecl* callee_param :
+       callee_function_decl.parameters()) {
+    clang::QualType param_type = callee_param->getType();
+    bool is_scalar_type = IsScalarType(ast_context, param_type);
+    if (!is_scalar_type) {
+      clang::QualType pointer_type = ast_context.getPointerType(param_type);
+      param_type = ast_context.getAttributedType(
+          clang::NullabilityKind::NonNull, pointer_type, pointer_type);
+    }
+    param_type_changed.push_back(!is_scalar_type);
+    thunk_param_types.push_back(param_type);
+  }
+
+  return {thunk_param_types, param_type_changed};
+}
+
+// Returns the thunk parameters using the callee function parameter identifiers.
+static auto BuildThunkParameters(
+    clang::ASTContext& ast_context,
+    const clang::FunctionDecl& callee_function_decl,
+    clang::FunctionDecl* thunk_function_decl)
+    -> llvm::SmallVector<clang::ParmVarDecl*> {
+  clang::SourceLocation clang_loc = callee_function_decl.getLocation();
+
+  unsigned num_params = thunk_function_decl->getNumParams();
+  CARBON_CHECK(callee_function_decl.getNumParams() == num_params);
+
+  const auto* thunk_function_proto_type =
+      thunk_function_decl->getFunctionType()->getAs<clang::FunctionProtoType>();
+
+  llvm::SmallVector<clang::ParmVarDecl*> thunk_params;
+  thunk_params.reserve(num_params);
+  for (unsigned i = 0; i < num_params; ++i) {
+    clang::ParmVarDecl* thunk_param = clang::ParmVarDecl::Create(
+        ast_context, thunk_function_decl, clang_loc, clang_loc,
+        callee_function_decl.getParamDecl(i)->getIdentifier(),
+        thunk_function_proto_type->getParamType(i), nullptr, clang::SC_None,
+        nullptr);
+    thunk_params.push_back(thunk_param);
+  }
+  return thunk_params;
+}
+
+// Returns the thunk function declaration given the callee function and the
+// thunk parameter types.
+static auto CreateThunkFunctionDecl(
+    clang::ASTContext& ast_context,
+    const clang::FunctionDecl& callee_function_decl,
+    llvm::ArrayRef<clang::QualType> thunk_param_types) -> clang::FunctionDecl* {
+  clang::SourceLocation clang_loc = callee_function_decl.getLocation();
+
+  clang::IdentifierInfo& identifier_info = ast_context.Idents.get(
+      callee_function_decl.getNameAsString() + "__carbon_thunk");
+
+  const clang::FunctionType* callee_function_type =
+      callee_function_decl.getFunctionType();
+  CARBON_CHECK(callee_function_type->isFunctionProtoType());
+
+  clang::QualType thunk_function_type = ast_context.getFunctionType(
+      callee_function_decl.getReturnType(), thunk_param_types,
+      callee_function_type->getAs<clang::FunctionProtoType>()
+          ->getExtProtoInfo());
+
+  clang::FunctionDecl* thunk_function_decl = clang::FunctionDecl::Create(
+      ast_context, ast_context.getTranslationUnitDecl(), clang_loc, clang_loc,
+      clang::DeclarationName(&identifier_info), thunk_function_type, nullptr,
+      clang::SC_Extern, false, false);
```

Comment:
I think we should emit these thunks with internal linkage.
```suggestion
      clang::SC_Static, false, false);
```

Reply:
Seems to break linkage.

```
$ bazel-bin/toolchain/carbon link hello_world.o main.o --output=demo
ld.lld: error: undefined symbol: hello_world(S) (.carbon_thunk)
>>> referenced by main.carbon:9
>>>               main.o:(main)
error: linker command failed with exit code 1 (use -v to see invocation)
```

Looking at the `thunk.carbon` test, it seems like the inline function is gone.
It replaces

```
// CHECK:STDOUT:
// CHECK:STDOUT: declare void @_Z3foos(i16)
// CHECK:STDOUT:
// CHECK:STDOUT: ; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite)
// CHECK:STDOUT: declare void @llvm.lifetime.start.p0(i64 immarg, ptr captures(none)) #0
// CHECK:STDOUT:
// CHECK:STDOUT: ; Function Attrs: alwaysinline mustprogress
// CHECK:STDOUT: define dso_local void @_Z3foos.carbon_thunk(ptr %x) #1 {
// CHECK:STDOUT: entry:
// CHECK:STDOUT:   %x.addr = alloca ptr, align 8
// CHECK:STDOUT:   store ptr %x, ptr %x.addr, align 8
// CHECK:STDOUT:   %0 = load ptr, ptr %x.addr, align 8
// CHECK:STDOUT:   %1 = load i16, ptr %0, align 2
// CHECK:STDOUT:   call void @_Z3foos(i16 signext %1)
// CHECK:STDOUT:   ret void
// CHECK:STDOUT: }
// CHECK:STDOUT:
// CHECK:STDOUT: attributes #0 = { nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) }
// CHECK:STDOUT: attributes #1 = { alwaysinline mustprogress "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="0" "target-cpu"="x86-64" "target-features"="+cmov,+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
```

with

```
// CHECK:STDOUT:
// CHECK:STDOUT: declare void @_Z3foos.carbon_thunk(ptr)
// CHECK:STDOUT:
// CHECK:STDOUT: declare void @_Z3foos(i16)
// CHECK:STDOUT:
// CHECK:STDOUT: ; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite)
// CHECK:STDOUT: declare void @llvm.lifetime.start.p0(i64 immarg, ptr captures(none)) #0
// CHECK:STDOUT:
// CHECK:STDOUT: attributes #0 = { nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) }
```

Reply:
Hm. I guess this is tied up in how we mark functions on the Clang side as referenced from the Carbon side. Maybe just add a TODO for now then, that the thunk should not have external linkage?

Reply:
Done.